base/iterators.jl

```diff
@@ -1430,4 +1430,57 @@ only(x::NamedTuple) = throw(
 
 Base.intersect(a::ProductIterator, b::ProductIterator) = ProductIterator(intersect.(a.iterators, b.iterators))
 
+"""
+    Iterator(f, initialstate)
+
+Iterable object with iterations defined by the function `f(state)`. The function takes the current state at each iteration, and follows the same
+rules as `iterate`. It must return either `(newvalue, newstate)` or `nothing`, in which case the iterator ends.
+
+See also: [`iterate`](@ref)
+
+# Examples
+
+```jldoctest
+julia> fib = Iterator((1,1)) do (a,b)
+           a, (b, a+b)
+       end;
+
+julia> reduce(hcat, Iterators.take(fib, 7))
+1×7 Matrix{Int64}:
+ 1  1  2  3  5  8  13
+
+julia> frac(c, z=0.0im) = Iterator{ComplexF64}((c, z)) do (c, z)
+           if real(z * z') < 4
+               z, (c, z^2 + c)
+           else
+               nothing
+           end
+       end;
+
+julia> [count(Returns(true), frac(-0.835-0.2321im, (k+j*im)/6)) for j in -4:4, k in -8:8]
+9×17 Matrix{Int64}:
+  2   2   2   3   3   3   5  41   8   4   3   3   2   2   2   2   1
+  2   3   5   4   5   8  20  11  17  23   4   3   3   3   2   2   2
+  4  10  17  12   7  56  18  58  33  22   6   5   4   5   4   3   2
+ 26  56  15  13  18  23  13  14  27  46   8   9  16  12   8   4   3
+ 10   7  62  17  16  23  11  12  39  12  11  23  16  17  62   7  10
+  3   4   8  12  16   9   8  46  27  14  13  23  18  13  15  56  26
+  2   3   4   5   4   5   6  22  33  58  18  56   7  12  17  10   4
+  2   2   2   3   3   3   4  23  17  11  20   8   5   4   5   3   2
+  1   2   2   2   2   3   3   4   8  41   5   3   3   3   2   2   2
+```
+"""
+struct Iterator{FuncType, Eltype, S}
+    f::FuncType
+    initialstate::S
+    Iterator{Eltype}(f::FuncType, initialstate::S) where {Eltype, FuncType, S} = new{FuncType, Eltype, S}(f, initialstate)
```

Comment:
I suggest satisfying the invariance for constructor  `T(args...)::T` for any type `T`. Please also limit the line length to 92 https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#code-formatting-guidelines

```suggestion
struct Iterator{Eltype, FuncType, S}
    f::FuncType
    initialstate::S
    Iterator{Eltype}(f::FuncType, initialstate::S) where {Eltype, FuncType, S} =
        new{Eltype, FuncType, S}(f, initialstate)

```